[REG-1872] add validation rules + update validations in general (#79)
Regression-Games/RegressionDocs: docs/validations/validations-rules-reference.mdx

---
sidebar_label: 'Rules Reference'
---

# Rules Reference

Rules can be asserted when running a Validation Suite against a gameplay recording, to ensure that the recording meets specific criteria. We provide a set of rule types that can be used to construct assertions based on various performance metrics. For each tick in the gameplay recording, all Rules are evaluated. If a Rule is violated the Validation Run will display a warning for that Rule.

## Rule Types

### Frames Per Second (with percentile)

Assert that the frames per second of the gameplay recording are above a threshold, within a specified percentile.

### Average Frames Per Second

> Assert that the average frames per second of the gameplay recording is above a threshold. This average is computed

Reviewer: ```suggestion
Assert that the average frames per second of the gameplay recording does not fall below a threshold. This average is computed
```

across all frames in a gameplay recording.

### Maximum Frame Time

> Assert that the maximum frame time of the gameplay recording is below a threshold. Frame time is the amount of time spent calculating and rendering a frame, and is measured in milliseconds. If a single frame exceeds the specified threshold, the rule will be violated.

Reviewer: ```suggestion
Assert that the maximum frame time of the gameplay recording does not exceed a threshold. Frame time is the amount of time spent calculating and rendering a frame, and is measured in milliseconds. If a single frame exceeds the specified threshold, the rule will be violated.
```

---
sidebar_label: 'Rules Reference'
---

# Rules Reference

> Rules can be asserted against a gameplay recording to ensure that the recording meets specific criteria. We provide a set of rule types that can be used to construct assertions based on various performance metrics.

Reviewer: I think we're missing an answer to the question "How is this different than a validation scenario?". Probably want to reword this to include that rules are applied to every tick in the gameplay session and act more as monitoring than a validation.

We'll also want to clarify somewhere that rules produce warnings, not errors.

Author: good call, added a bit about this

Rules belong to a Validation Suite, and are asserted each time the Validation Suite is run.

## Rule Types

### Frames Per Second (with percentile)

> Assert that the frames per second of the gameplay recording are above a threshold, within a specified percentile.

Reviewer: These descriptions need to account for the fact that meeting the threshold does not violate it. 

"are above a threshold" -> "do not fall below the threshold"
similar for the other rule descriptions

Author: I agree. I have the 3 descriptions containing either "above a threshold", or "below a threshold"

Reviewer: ```suggestion
Assert that the frames per second of the gameplay recording do not fall below a threshold, within a specified percentile.
```

Author: ah there was some confusion on my part regarding your original comment. Changes made

Author: f

### Average Frames Per Second

> Assert that the average frames per second of the gameplay recording is above a threshold.

Reviewer 2: I think we need to clarify how this average is calculated. i.e. whether this is an average over the entire gameplay recording or whether this is a rolling average over the last N ticks/frames.

Author: thanks. Added

### Maximum Frame Time

> Assert that the maximum frame time of the gameplay recording is below a threshold.

Reviewer 2: Readers will likely be confused about what frame time is. We should clarify that this represents the time spent calculating and rendering a frame and is the inverse of FPS.

Author: good idea added
